// create-plasmic-app/dist/templates/gatsby.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapAppRootForCodegen = exports.makeGatsbyPlasmicInit = exports.GATSBY_SSR_CONFIG = exports.makeGatsbyHostPage = exports.GATSBY_PLUGIN_CONFIG = exports.GATSBY_404 = exports.makeGatsbyDefaultPage = void 0;
const file_utils_1 = require("../utils/file-utils");
const makeGatsbyDefaultPage = (format) => {
    const ts = format === "ts";
    return `
import React from "react";
import Helmet from "react-helmet";
import {
  PlasmicComponent,
  PlasmicRootProvider,${file_utils_1.ifTs(ts, `
  InitOptions,
  ComponentRenderData,`)}
} from "@plasmicapp/loader-gatsby";
import { graphql } from "gatsby";
import { initPlasmicLoaderWithRegistrations } from "../plasmic-init";

export const query = graphql\`
  query ($path: String) {
    plasmicComponents(componentNames: [$path])
    plasmicOptions
  }
\`;

${file_utils_1.ifTs(ts, `interface PlasmicGatsbyPageProps {
  data: {
    plasmicOptions: InitOptions
    plasmicComponents: ComponentRenderData
  }
}
`)}
const PlasmicGatsbyPage = ({ data }${file_utils_1.ifTs(ts, ": PlasmicGatsbyPageProps")}) => {
  const {
    plasmicComponents,
    plasmicOptions,
  } = data;
  const pageMeta = plasmicComponents.entryCompMetas[0];
  const pageMetadata = pageMeta.pageMetadata;
  return (
    <PlasmicRootProvider
      loader={initPlasmicLoaderWithRegistrations(plasmicOptions)}
      prefetchedData={plasmicComponents}
    >
      <Helmet>
        {pageMetadata.title && <title>{pageMetadata.title}</title>}
        {pageMetadata.title && <meta property="og:title" content={pageMetadata.title} /> }
        {pageMetadata.description && <meta property="og:description" content={pageMetadata.description} />}
        {pageMetadata.openGraphImageUrl && <meta property="og:image" content={pageMetadata.openGraphImageUrl} />}
      </Helmet>
      <PlasmicComponent component={pageMeta.name} />
    </PlasmicRootProvider>
  );
};

export default PlasmicGatsbyPage;
`.trim();
};
exports.makeGatsbyDefaultPage = makeGatsbyDefaultPage;
exports.GATSBY_404 = `
const NotFound = () => {
  return "Not Found";
};
export default NotFound;
`.trim();
const GATSBY_PLUGIN_CONFIG = (projectId, projectApiToken, useTypescript) => `
{
  resolve: "@plasmicapp/loader-gatsby",
  options: {
    projects: [
      {
        id: "${projectId}",
        token: "${projectApiToken}",
      },
    ], // An array of project ids.
    defaultPlasmicPage: require.resolve("./src/templates/defaultPlasmicPage.${useTypescript ? "tsx" : "jsx"}"),
  },
},
`;
exports.GATSBY_PLUGIN_CONFIG = GATSBY_PLUGIN_CONFIG;
const makeGatsbyHostPage = (opts) => {
    const { useTypescript, scheme } = opts;
    if (scheme === "loader") {
        return `
import * as React from "react"
import {
  PlasmicCanvasHost${file_utils_1.ifTs(useTypescript, `, InitOptions`)}
} from "@plasmicapp/loader-gatsby"
import { graphql } from "gatsby"
import { initPlasmicLoaderWithRegistrations } from "../plasmic-init"

export const query = graphql\`
  query {
    plasmicOptions
  }
\`

${file_utils_1.ifTs(useTypescript, `interface HostProps {
  data: {
    plasmicOptions: InitOptions;
  }
}
`)}
export default function Host({ data }${file_utils_1.ifTs(useTypescript, ": HostProps")}) {
  const { plasmicOptions } = data
  initPlasmicLoaderWithRegistrations(plasmicOptions)
  return <PlasmicCanvasHost />
}
    `.trim();
    }
    else {
        return `
import * as React from "react"
import { PlasmicCanvasHost, registerComponent } from "@plasmicapp/host";

// You can register any code components that you want to use here; see
// https://docs.plasmic.app/learn/code-components-ref/
// And configure your Plasmic project to use the host url pointing at
// the /plasmic-host page of your nextjs app (for example,
// http://localhost:3000/plasmic-host).  See
// https://docs.plasmic.app/learn/app-hosting/#set-a-plasmic-project-to-use-your-app-host

// registerComponent(...)

export default function PlasmicHost() {
  return (
    <PlasmicCanvasHost />
  );
}
    `;
    }
};
exports.makeGatsbyHostPage = makeGatsbyHostPage;
exports.GATSBY_SSR_CONFIG = `
/**
 * Implement Gatsby's SSR (Server Side Rendering) APIs in this file.
 *
 * See: https://www.gatsbyjs.com/docs/ssr-apis/
 */

const React = require("react")

/**
 * Add preamble to allow functional code components in studio.
 *
 * See: https://docs.plasmic.app/learn/functional-code-components/
 */
const HeadComponents = [
  <script
    key="plasmic-preamble"
    src="https://static1.plasmic.app/preamble.js"
  />,
  <script
    key="plasmic-hmr"
    type="text/javascript"
    dangerouslySetInnerHTML={{
      __html: \`
        if (typeof window !== "undefined" && /\\\\/plasmic-host\\\\/?$/.test(window.location.pathname)) {
          const RealEventSource = window.EventSource;
          window.EventSource = function(url, config) {
            if (/[^a-zA-Z]hmr($|[^a-zA-Z])/.test(url)) {
              console.warn("Plasmic: disabled EventSource request for", url);
              return {
                onerror() {}, onmessage() {}, onopen() {}, close() {}
              };
            } else {
              return new RealEventSource(url, config);
            }
          }
        }
      \`,
    }}
  />
]

const isProduction = process.env.NODE_ENV === "production"

exports.onRenderBody = ({ pathname, setHeadComponents }) => {
  /**
   * We add the preamble tag script to all pages during development mode
   * because during development all pages are dynamically rendered based
   * on \`/\` route, during production we add it only in \`/plasmic-host/\`
   */
  if (!isProduction || pathname === "/plasmic-host/") {
    setHeadComponents(HeadComponents)
  }
}
`.trim();
const makeGatsbyPlasmicInit = (format) => {
    const ts = format === "ts";
    return `
import {
  initPlasmicLoader,${file_utils_1.ifTs(ts, `
  InitOptions,`)}
} from "@plasmicapp/loader-gatsby";

export function initPlasmicLoaderWithRegistrations(plasmicOptions${file_utils_1.ifTs(ts, ": InitOptions")}) {
  const PLASMIC = initPlasmicLoader(plasmicOptions);

  // You can register any code components that you want to use here; see
  // https://docs.plasmic.app/learn/code-components-ref/
  // And configure your Plasmic project to use the host url pointing at
  // the /plasmic-host page of your nextjs app (for example,
  // http://localhost:8000/plasmic-host).  See
  // https://docs.plasmic.app/learn/app-hosting/#set-a-plasmic-project-to-use-your-app-host

  // PLASMIC.registerComponent(...);

  return PLASMIC;
}
`.trim();
};
exports.makeGatsbyPlasmicInit = makeGatsbyPlasmicInit;
function wrapAppRootForCodegen() {
    return `
import React from "react";
import { PlasmicRootProvider } from "@plasmicapp/react-web";

export const wrapRootElement = ({ element }) => {
  return (
    <PlasmicRootProvider>
      {element}
    </PlasmicRootProvider>
  );
}
  `;
}
exports.wrapAppRootForCodegen = wrapAppRootForCodegen;
