// @plasmicapp/cli/dist/utils/semver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.outside = exports.ltr = exports.gtr = exports.coerce = exports.maxSatisfying = exports.minSatisfying = exports.sortDesc = exports.sortAsc = exports.neq = exports.lte = exports.gte = exports.toCaretRange = exports.toTildeRange = exports.satisfies = exports.validRange = exports.lt = exports.gt = exports.eq = exports.patch = exports.minor = exports.major = exports.prerelease = exports.inc = exports.valid = exports.isLatest = exports.latestTag = void 0;
const lodash_1 = __importDefault(require("lodash"));
const semverlib = __importStar(require("semver"));
exports.latestTag = "latest";
exports.isLatest = (v) => v === exports.latestTag;
exports.valid = (v) => exports.isLatest(v) ? exports.latestTag : semverlib.valid(v);
exports.inc = (v, release) => exports.isLatest(v) ? exports.latestTag : semverlib.inc(v, release);
exports.prerelease = (v) => exports.isLatest(v) ? [] : semverlib.prerelease(v);
exports.major = (v) => exports.isLatest(v) ? exports.latestTag : semverlib.major(v);
exports.minor = (v) => exports.isLatest(v) ? exports.latestTag : semverlib.minor(v);
exports.patch = (v) => exports.isLatest(v) ? exports.latestTag : semverlib.patch(v);
exports.eq = (v1, v2) => (exports.isLatest(v1) && exports.isLatest(v2)) ||
    (!exports.isLatest(v1) && !exports.isLatest(v2) && semverlib.eq(v1, v2));
exports.gt = (v1, v2) => (exports.isLatest(v1) && !exports.isLatest(v2)) ||
    (!exports.isLatest(v1) && !exports.isLatest(v2) && semverlib.gt(v1, v2));
exports.lt = (v1, v2) => (!exports.isLatest(v1) && exports.isLatest(v2)) ||
    (!exports.isLatest(v1) && !exports.isLatest(v2) && semverlib.lt(v1, v2));
exports.validRange = (range) => exports.isLatest(range) ? exports.latestTag : semverlib.validRange(range);
exports.satisfies = (v, range) => (exports.isLatest(range) && !!exports.valid(v)) ||
    (!exports.isLatest(v) && !exports.isLatest(range) && semverlib.satisfies(v, range));
exports.toTildeRange = (v) => exports.isLatest(v) ? exports.latestTag : !!semverlib.valid(v) ? "~" + v : null;
exports.toCaretRange = (v) => exports.isLatest(v) ? exports.latestTag : !!semverlib.valid(v) ? "^" + v : null;
exports.gte = (v1, v2) => exports.eq(v1, v2) || exports.gt(v1, v2);
exports.lte = (v1, v2) => exports.eq(v1, v2) || exports.lt(v1, v2);
exports.neq = (v1, v2) => !exports.eq(v1, v2);
exports.sortAsc = (versions) => lodash_1.default.cloneDeep(versions).sort((v1, v2) => exports.gt(v1, v2) ? +1 : exports.eq(v1, v2) ? 0 : -1);
exports.sortDesc = (versions) => exports.sortAsc(versions).reverse();
exports.minSatisfying = (versions, range) => { var _a; return (_a = exports.sortAsc(versions).find((v) => exports.satisfies(v, range))) !== null && _a !== void 0 ? _a : null; };
exports.maxSatisfying = (versions, range) => { var _a; return (_a = exports.sortDesc(versions).find((v) => exports.satisfies(v, range))) !== null && _a !== void 0 ? _a : null; };
exports.coerce = (v) => { var _a; return exports.isLatest(v) ? exports.latestTag : (_a = semverlib.coerce(v)) === null || _a === void 0 ? void 0 : _a.version; };
exports.gtr = (version, range) => (exports.isLatest(version) && !exports.isLatest(range)) ||
    (!exports.isLatest(version) && !exports.isLatest(range) && semverlib.gtr(version, range));
exports.ltr = (version, range) => (!exports.isLatest(version) && exports.isLatest(range)) ||
    (!exports.isLatest(version) && !exports.isLatest(range) && semverlib.ltr(version, range));
exports.outside = (version, range, hilo) => (hilo === ">" && exports.gtr(version, range)) ||
    (hilo === "<" && exports.ltr(version, range)) ||
    (!hilo && (exports.gtr(version, range) || exports.ltr(version, range)));
