// @plasmicapp/cli/dist/lib.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var auth_1 = require("./actions/auth");
Object.defineProperty(exports, "auth", { enumerable: true, get: function () { return auth_1.auth; } });
var fix_imports_1 = require("./actions/fix-imports");
Object.defineProperty(exports, "fixImports", { enumerable: true, get: function () { return fix_imports_1.fixImports; } });
var init_1 = require("./actions/init");
Object.defineProperty(exports, "initPlasmic", { enumerable: true, get: function () { return init_1.initPlasmic; } });
var project_token_1 = require("./actions/project-token");
Object.defineProperty(exports, "getProjectApiToken", { enumerable: true, get: function () { return project_token_1.getProjectApiToken; } });
var sync_1 = require("./actions/sync");
Object.defineProperty(exports, "sync", { enumerable: true, get: function () { return sync_1.sync; } });
var upload_bundle_1 = require("./actions/upload-bundle");
Object.defineProperty(exports, "uploadJsBundle", { enumerable: true, get: function () { return upload_bundle_1.uploadJsBundle; } });
var watch_1 = require("./actions/watch");
Object.defineProperty(exports, "watchProjects", { enumerable: true, get: function () { return watch_1.watchProjects; } });
var deps_1 = require("./deps");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return deps_1.logger; } });
var error_1 = require("./utils/error");
Object.defineProperty(exports, "HandledError", { enumerable: true, get: function () { return error_1.HandledError; } });
Object.defineProperty(exports, "handleError", { enumerable: true, get: function () { return error_1.handleError; } });
var get_context_1 = require("./utils/get-context");
Object.defineProperty(exports, "setMetadataEnv", { enumerable: true, get: function () { return get_context_1.setMetadataEnv; } });
